keps/sig-node/5334-image-pull-progrss/README.md

```diff
@@ -0,0 +1,809 @@
+
+### User Stories (Optional)
+
+#### Story 1
+
+A user deploys an application to a Kubernetes cluster. As the Pod is scheduled and created,
+the only event visible in the Pod's event list is "Pulling image,"
+with no indication of progress or estimated time to completion.
+
+#### Story 2
+
+Allow kubectl and crictl to see image pull progress like other container tools.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+TODO
+
+## Design Details
+
+This feature will adopt the established architectural patterns used by existing Kubernetes streaming features such as exec, port-forward, and logs.
+
+Proposed changes to the cri:
+
+This API will return a URL, similar to the existing exec and port-forward interfaces, provided by the CRI implementation.
+
+``` proto
+// Runtime service defines the public APIs for remote container runtimes
+service RuntimeService {
+    ...
+
+    // ImagePullProgress prepares a streaming endpoint to image pull from a PodSandbox.
+    rpc ImagePullProgress(ImagePullProgressRequest) returns (ImagePullProgressResponse) {}
+    
+    ...
+}
+
+message ImagePullProgressRequest {
+    // ID of the pod to which to progresss of the image pull.
+    string pod_sandbox_id = 1;
+}
+
+message ImagePullProgressResponse {
+    // Fully qualified URL of the image pull progresss streaming server.
+    string url = 1;
+}
+
+```
+
+Data format pushed by the cri-side:
+
+CRI implementations may report image pull progress at different levels of detail: either for the entire image or for individual layers. 
+
+- When reporting progress for the entire image, the `Name` field should be set to the image name.
+- When reporting progress per layer, the `Name` field should be set to the layer's digest.
+
+Implementers can choose either approach.
+
+Implementers also determine the frequency of pushes.
+
+``` golang
+type Progress struct {
+	Name     string `json:"name"`
+	Progress int64  `json:"progress"`
+	Total    int64  `json:"total"`
+	Error    string `json:"error,omitempty"`
+}
+```
+
+Proposed changes to the crictl:
+
+- Add a new flag to `crictl pull` that allows users to monitor the image pull progress.
+
+
+Proposed changes to the kubelet:
+
+It is expected to be implemented after the CRI implementation supporting the CRI API is released.
+
+- Add a new route `/imagePullProgress/{podNamespace}/{podID}` to the Kubelet server.
```

Comment:
is camel case typical for http endpoints like this? it seems a bit odd to me but I'm not familiar with convention

Reply:
Yes, it is the typical camel case. I am referring to the portForward endpoint.

https://github.com/kubernetes/kubernetes/blob/fd53f7292c7d5899135fddd928c0dc3844126820/pkg/kubelet/server/server.go#L556